### PR TITLE
Use C# newer language features

### PR DESCRIPTION
**Pull request by srs6814**
Have tried to use the following language syntax/features

1) Use null propogation
2) Pattern based usings declaration
3) Merge sequential checks
4) Convert to method group instead of using lambda
5) Simplify ternary operations

### CODE REVIEW

File: src/GraphQL.Tests/Validation/ArgumentsOfCorrectTypeTests.cs [hunk 1]
```diff
@@ -887,10 +887,7 @@ public static class ValidationExtensions
             int column,
             IEnumerable<string> errors = null)
         {
-            if (errors == null)
-            {
-                errors = new[] { $"Expected type \"{typeName}\", found {value}." };
-            }
+            errors ??= new[] {$"Expected type \"{typeName}\", found {value}."};
```
Comment (sungam3r) [inline src/GraphQL.Tests/Validation/ArgumentsOfCorrectTypeTests.cs:890]:
```suggestion
            errors ??= new[] { $"Expected type \"{typeName}\", found {value}." };
```

File: src/GraphQL/Conversion/ValueConverter.cs [hunk 2]
```diff
@@ -49,9 +49,9 @@ static ValueConverter()
 
                 return Convert.ToBoolean(value, NumberFormatInfo.InvariantInfo).Boxed();
             });
-            Register<string, Guid>(value => Guid.Parse(value));
+            Register<string, Guid>(Guid.Parse);
```
Comment (sungam3r) [inline src/GraphQL/Conversion/ValueConverter.cs:52]: I would not want to change this signature to keep the overall style.

File: src/GraphQL/Conversion/ValueConverter.cs [hunk 3]
```diff
@@ -49,9 +49,9 @@ static ValueConverter()
...
             Register<string, Uri>(value => new Uri(value));
-            Register<string, byte[]>(value => Convert.FromBase64String(value)); // such a built-in conversion for string->byte[] seems useful
+            Register<string, byte[]>(Convert.FromBase64String); // such a built-in conversion for string->byte[] seems useful
```
Comment (sungam3r) [inline src/GraphQL/Conversion/ValueConverter.cs:54]: Same here.

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs [hunk 4]
```diff
@@ -113,7 +113,7 @@ private static string GetPropertyName<TSourceType>(Expression<Func<TSourceType,
             if (expression.Body is MemberExpression m1)
                 return m1.Member.Name;
 
-            if (expression.Body is UnaryExpression u && u.Operand is MemberExpression m2)
+            if (expression.Body is UnaryExpression {Operand: MemberExpression m2} u)
```
Comment (sungam3r) [inline src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:116]: I know this syntax is becoming fashionable for some developers. However, I would not want to use it, at least not yet. It is not easier to read than the current version. @Shane32  ?

File: src/GraphQL/ResolveFieldContext/ResolveFieldContextExtensions.cs [hunk 5]
```diff
@@ -61,7 +61,7 @@ private static bool TryGetArgument(this IResolveFieldContext context, Type argum
         /// <summary>Determines if the specified field argument has been provided in the GraphQL query request</summary>
         public static bool HasArgument(this IResolveFieldContext context, string name)
         {
-            var isIntrospection = context.ParentType == null ? context.FieldDefinition.IsIntrospectionField() : context.ParentType.IsIntrospectionType();
+            var isIntrospection = context.ParentType?.IsIntrospectionType() ?? context.FieldDefinition.IsIntrospectionField();
```
Comment (sungam3r) [inline src/GraphQL/ResolveFieldContext/ResolveFieldContextExtensions.cs:64]: This expression is not equivalent to the previous one.

File: src/GraphQL/Introspection/__Field.cs [hunk 6]
```diff
@@ -28,7 +28,7 @@ public __Field()
                 {
                     // We have to iterate over all schema types because FieldType has no reference to the GraphType to which it belongs.
                     var fieldOwner = context.Schema.AllTypes.OfType<IComplexGraphType>().Single(t => t.Fields.Contains(context.Source));
-                    if (fieldOwner is IImplementInterfaces implementation && implementation.ResolvedInterfaces != null)
+                    if (fieldOwner is IImplementInterfaces {ResolvedInterfaces: { }} implementation)
```
Comment (sungam3r) [inline src/GraphQL/Introspection/__Field.cs:31]: It's harder to understand. Please revert.

File: src/GraphQL/Types/Relay/ConnectionType.cs [hunk 7]
```diff
@@ -16,9 +16,9 @@ public class ConnectionType<TNodeType, TEdgeType> : ObjectGraphType<object>
         /// <inheritdoc/>
         public ConnectionType()
         {
-            Name = string.Format("{0}Connection", typeof(TNodeType).GraphQLName());
-            Description = string.Format("A connection from an object to a list of objects of type `{0}`.",
-                typeof(TNodeType).GraphQLName());
+            Name = $"{typeof(TNodeType).GraphQLName()}Connection";
```
Comment (sungam3r) [inline src/GraphQL/Types/Relay/ConnectionType.cs:19]: Move `typeof(TNodeType).GraphQLName()` into local variable.

File: src/GraphQL/Types/Relay/EdgeType.cs [hunk 8]
```diff
@@ -14,10 +14,9 @@ public class EdgeType<TNodeType> : ObjectGraphType<object>
         /// <inheritdoc/>
         public EdgeType()
         {
-            Name = string.Format("{0}Edge", typeof(TNodeType).GraphQLName());
-            Description = string.Format(
-                "An edge in a connection from an object to another object of type `{0}`.",
-                typeof(TNodeType).GraphQLName());
+            Name = $"{typeof(TNodeType).GraphQLName()}Edge";
```
Comment (sungam3r) [inline src/GraphQL/Types/Relay/EdgeType.cs:17]: Move `typeof(TNodeType).GraphQLName()` into local variable.

File: src/GraphQL/Utilities/Federation/FederatedSchemaBuilder.cs [hunk 9]
```diff
@@ -121,7 +121,7 @@ private bool FindSelectionToAmend(SelectionSet selectionSet, Document document,
         {
             foreach (var selection in selectionSet.Selections)
             {
-                if (selection is Field childField && childField.Name == "__typename")
+                if (selection is Field {Name: "__typename"} childField)
```
Comment (sungam3r) [inline src/GraphQL/Utilities/Federation/FederatedSchemaBuilder.cs:124]: Please revert changes in this file.

**Comment by sungam3r** [comment]
@srs6814 I reviewed almost all the files and only when I came to the very end I noticed that your PR is targeting the `master`. We are preparing a v4 release and all the work is happening in the `develop` branch. Please change your target branch.

File: src/GraphQL/Conversion/ValueConverter.cs (hunk 2 above)
Comment (Shane32) [reply src/GraphQL/Conversion/ValueConverter.cs:52]: Agreed

File: src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs (hunk 4 above)
Comment (Shane32) [reply src/GraphQL/Types/Composite/AutoRegisteringObjectGraphType.cs:116]: Agreed

File: src/GraphQL/Introspection/__Field.cs (hunk 6 above)
Comment (Shane32) [reply src/GraphQL/Introspection/__Field.cs:31]: Agreed

File: src/GraphQL/ResolveFieldContext/ResolveFieldContextExtensions.cs (hunk 5 above)
Comment (Shane32) [reply src/GraphQL/ResolveFieldContext/ResolveFieldContextExtensions.cs:64]: > @srs6814 I reviewed almost all the files and only when I came to the very end I noticed that your PR is targeting the `master`. We are preparing a v4 release and all the work is happening in the `develop` branch. Please change your target branch.

Agreed.

File: src/GraphQL/Utilities/SchemaPrinter.cs [hunk 10]
```diff
@@ -203,7 +203,7 @@ public string PrintEnum(EnumerationGraphType type)
 
         public string PrintInputObject(IInputObjectGraphType type)
         {
-            var fields = type.Fields.Select(x => PrintInputValue(x));
+            var fields = type.Fields.Select(PrintInputValue);
```
Comment (sungam3r) [inline src/GraphQL/Utilities/SchemaPrinter.cs:206]: Please revert. Also see https://github.com/dotnet/roslyn/issues/39869

File: src/GraphQL/Types/Relay/ConnectionType.cs [hunk 11]
```diff
@@ -16,9 +16,10 @@ public class ConnectionType<TNodeType, TEdgeType> : ObjectGraphType<object>
         /// <inheritdoc/>
         public ConnectionType()
         {
-            Name = string.Format("{0}Connection", typeof(TNodeType).GraphQLName());
-            Description = string.Format("A connection from an object to a list of objects of type `{0}`.",
-                typeof(TNodeType).GraphQLName());
+            string graphQLTypeName = typeof(TNodeType).GraphQLName();
+            Name = $"{graphQLTypeName}Connection";
+            Description =
+                $"A connection from an object to a list of objects of type `{typeof(TNodeType).GraphQLName()}`.";
```
Comment (sungam3r) [inline src/GraphQL/Types/Relay/ConnectionType.cs:22]:
```suggestion
                $"A connection from an object to a list of objects of type `{graphQLTypeName}`.";
```

File: src/GraphQL/Types/Relay/EdgeType.cs [hunk 12]
```diff
@@ -14,10 +14,10 @@ public class EdgeType<TNodeType> : ObjectGraphType<object>
...
         public EdgeType()
         {
-            Name = string.Format("{0}Edge", typeof(TNodeType).GraphQLName());
-            Description = string.Format(
-                "An edge in a connection from an object to another object of type `{0}`.",
-                typeof(TNodeType).GraphQLName());
+            string graphQLTypeName = typeof(TNodeType).GraphQLName();
+            Name = $"{graphQLTypeName}Edge";
+            Description =
+                $"An edge in a connection from an object to another object of type `{typeof(TNodeType).GraphQLName()}`.";
```
Comment (sungam3r) [inline src/GraphQL/Types/Relay/EdgeType.cs:20]:
```suggestion
                $"An edge in a connection from an object to another object of type `{graphQLTypeName}`.";
```

**Comment by srs6814** [comment]
> @srs6814 I reviewed almost all the files and only when I came to the very end I noticed that your PR is targeting the `master`. We are preparing a v4 release and all the work is happening in the `develop` branch. Please change your target branch.

Thank you @sungam3r for your review and helpful comments.

File: src/GraphQL/ResolveFieldContext/ResolveFieldContextExtensions.cs (hunk 5 above)
Comment (srs6814) [reply src/GraphQL/ResolveFieldContext/ResolveFieldContextExtensions.cs:65]: @sungam3r 
isn't it similar to the previous one?

if context.ParentType == null, it fallbacks to context.FieldDefinition.IsIntrospectionField

Comment (sungam3r) [reply src/GraphQL/ResolveFieldContext/ResolveFieldContextExtensions.cs:65]: `context.ParentType` may be not null, but `context.ParentType.IsIntrospectionType()` still _may be_ null. There is no fallback in the  first case.

File: src/GraphQL/Conversion/ValueConverter.cs [hunk 13]
```diff
@@ -51,7 +51,7 @@ static ValueConverter()
             });
             Register<string, Guid>(value => Guid.Parse(value));
             Register<string, Uri>(value => new Uri(value));
-            Register<string, byte[]>(value => Convert.FromBase64String(value)); // such a built-in conversion for string->byte[] seems useful
+            Register<string, byte[]>(Convert.FromBase64String); // such a built-in conversion for string->byte[] seems useful
```
Comment (sungam3r) [inline src/GraphQL/Conversion/ValueConverter.cs:54]:
```suggestion
            Register<string, byte[]>(value => Convert.FromBase64String(value)); // such a built-in conversion for string->byte[] seems useful
```

**Comment by sungam3r** [comment]
@Shane32 OK to merge? I fixed compilation errors.

**Review by Shane32 — APPROVED** [review]

**Pull request merged**